### ISSUE TITLE
Introduce absurd patterns instead of `impossible` clauses as we have now

### ISSUE THREAD

**Issue by ahmadsalim** [idris-lang/Idris-dev#3137]
Issue #2347 had some discussion about replacing `impossible` clauses with absurd patterns.
I will copy the example from @david-christiansen below:

> So one can implement void like this:
> 
> ```idris
> void : Void -> a
> void 💩
> ```
> But absurd patterns also dig down, explaining precisely where the unmatchable bit is:
> 
> ```
> data Garbage = Bin Void
> 
> empty : Garbage -> a
> empty (Bin 💩)
> ```
> or
> 
> ```
> toUnit : Fin 1 -> ()
> toUnit FZ = ()
> toUnit (FS 💩)
> ```
>

**Comment by milessabin** [comment]
I'm not saying it's a bad thing, or that you shouldn't do it, but this is how that symbol renders in Gmail,

![](https://pbs.twimg.com/media/ChXIJIMWwAApXMQ.jpg)

**Comment by edwinb** [comment]
Perhaps this will happen if someone actually steps up and does it - I'm not going to, though, at least not in the near future.

**Comment by ahmadsalim** [comment]
@edwinb The idea was just that this particular feature request was buried inside the conversation in the other issue, and I thought I would make a separate issue for it (while cleaning up). Although, hopefully we do get some contributors for this,

**Comment by xekoukou** [comment]
Don't we have `absurd` already?

**Comment by ahmadsalim** [comment]
@xekoukou  `absurd` is an interface method which you have to implement and call explicitly, whereas this is a proposal for a native feature where the absurdness is explicitly pointed out in patterns.

**Comment by xekoukou** [comment]
I agree. What I do not know is whether Void or Uninhabited would be needed with this change. I would propose to remove them.

**Comment by ahmadsalim** [comment]
@xekoukou This is only a feature proposal at this stage.
I would guess that `Uninhabited` would be less useful to have if we have absurd patterns, but still could be used for some things. `Void` is however essential to be able to express things like negation, etc.

**Comment by Melvar** [comment]
I am generally in favor of adding such a thing, because it seems more flexible.
